### DISCUSSION TITLE
Yawning-Titan install failing due to Gym not compatible with setuptools>=67

### DISCUSSION THREAD

**Discussion by ChrisMcCarthyDev** [dstl/YAWNING-TITAN#44] [General]
We are aware of a dependency issue that is causing the Yawning-Titan install to fail with the following error:

``` 
(venv) PS C:\Users\ChristopherMcCarthy\source\azure_devops\AIDT\YAWNING-TITAN> pip install -e .[dev]    
Obtaining file:///C:/Users/ChristopherMcCarthy/source/azure_devops/AIDT/YAWNING-TITAN
  Preparing metadata (setup.py) ... done
Collecting ray[rllib]@ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-3.0.0.dev0-cp310-cp310-win_amd64.whl
  Downloading https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-3.0.0.dev0-cp310-cp310-win_amd64.whl (21.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.6/21.6 MB 5.0 MB/s eta 0:00:00
Collecting dm-tree==0.1.7
  Downloading dm_tree-0.1.7-cp310-cp310-win_amd64.whl (90 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.9/90.9 kB 2.6 MB/s eta 0:00:00
Collecting gym==0.21.0
  Downloading gym-0.21.0.tar.gz (1.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 7.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [1 lines of output]
      error in gym setup command: 'extras_require' must be a dictionary whose values are strings or lists of strings containing valid project/versio
n requirement specifiers.
× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

This error is due to [setuptools](https://pypi.org/project/setuptools/) version >= 67. For further details, see [[Bug Report] Gym==0.21 not installing anymore](https://github.com/openai/gym/issues/3200).

If you are experiencing this same issue, downgrading setuptools to version 66 should fix it:
``` bash
pip install --upgrade setuptools==66
```

**Comment by jackdmoody** [comment]
Hello! I have tried this and I am still receiving the same error. Do you have any recommendations?

**Reply by ChrisMcCarthyDev** [reply]
Hi @jackdmoody0427. It seems there's been a few things happen around the same time that are causing `pip install gym==0.21.0` to fail. The first is the setuptools upgrade, the second is pip version >=23.1 no longer supporting legacy install using setup.py. 

See https://github.com/pypa/pip/issues/8368 for more details.

The following explains the gradual deprecation of the legacy setup.py.

> The pyproject.toml file was introduced in [PEP-518](https://www.python.org/dev/peps/pep-0518/) (2016) as a way of separating configuration of the build system from a specific, optional library (setuptools) and also enabling setuptools to install  itself without already being installed. Subsequently [PEP-621](https://www.python.org/dev/peps/pep-0621/) (2020) introduces the idea that the pyproject.toml file be used for wider project configuration and [PEP-660](https://www.python.org/dev/peps/pep-0660/) (2021) proposes finally doing away with the need for setup.py for editable installation using pip.

**Source: [Understanding setup.py, setup.cfg and pyproject.toml in Python](https://ianhopkinson.org.uk/2022/02/understanding-setup-py-setup-cfg-and-pyproject-toml-in-python/)**

We have an internal build pipeline which is currently fixing on specific versions of 'pip', 'setuptools', and 'wheel'. Try running the below in your Yawning-Titan virtual environment (or main system Python if you're not using a virtual environment).

  ``` bash
python -m pip install --upgrade pip==23.0.1
pip install wheel==0.38.4 --upgrade
pip install setuptools==66 --upgrade
```

If this still doesn't work, please run the following script and paste the output along with a copy of the error message you're receiving when attempting to install Yawning-Titan.

``` bash
python -c 'import platform; print(platform.platform())'
python --version
pip show pip
pip show wheel
pip show setuptools
```
